File: BackEnd/PHP/laravelshop/app/Http/Controllers/Api/V1/Auth/AuthController.php
<?php

namespace App\Http\Controllers\Api\V1\Auth;

use App\Http\Controllers\Api\V1\ApiController;
use App\Modules\Auth\Requests\LoginRequest;
use App\Modules\Auth\Requests\PasswordResetConfirmRequest;
use App\Modules\Auth\Requests\PasswordResetRequestRequest;
use App\Modules\Auth\Requests\PasswordResetVerifyRequest;
use App\Modules\Auth\Requests\RefreshTokenRequest;
use App\Modules\Auth\Requests\ResendVerificationRequest;
use App\Modules\Auth\Requests\SignupRequest;
use App\Modules\Auth\Requests\SocialCompleteRequest;
use App\Modules\Auth\Requests\SocialLinkRequest;
use App\Modules\Auth\Requests\VerifyEmailRequest;
use App\Modules\Auth\Services\AuthService;
use Illuminate\Http\Request;

class AuthController extends ApiController
{
    public function __construct(
        private readonly AuthService $authService,
    ) {}

    public function signup(SignupRequest $request)
    {
        return $this->success($this->authService->signup($request->validated()), status: 201);
    }

    public function verifyEmail(VerifyEmailRequest $request)
    {
        return $this->success($this->authService->verifyEmail(
            $request->string('email')->value(),
            $request->string('code')->value(),
        ));
    }

    public function resendVerification(ResendVerificationRequest $request)
    {
        return $this->success($this->authService->resendVerification($request->string('email')->value()));
    }

    public function login(LoginRequest $request)
    {
        return $this->success($this->authService->login(
            $request->string('email')->value(),
            $request->string('password')->value(),
        ));
    }

    public function logout(Request $request)
    {
        return $this->success($this->authService->logout($request->user()));
    }

    public function refresh(RefreshTokenRequest $request)
    {
        return $this->success($this->authService->refresh($request->string('refreshToken')->value()));
    }

    public function requestPasswordReset(PasswordResetRequestRequest $request)
    {
        return $this->success($this->authService->requestPasswordReset(
            $request->string('email')->value(),
            $request->string('phone')->value(),
        ));
    }

    public function verifyPasswordResetCode(PasswordResetVerifyRequest $request)
    {
        return $this->success($this->authService->verifyPasswordResetCode(
            $request->string('email')->value(),
            $request->string('code')->value(),
        ));
    }

    public function confirmPasswordReset(PasswordResetConfirmRequest $request)
    {
        return $this->success($this->authService->confirmPasswordReset(
            $request->string('resetToken')->value(),
            $request->string('newPassword')->value(),
        ));
    }

    public function socialRedirect(string $provider)
    {
        return redirect()->away($this->authService->socialRedirect($provider));
    }

    public function socialCallback(Request $request, string $provider)
    {
        return $this->success($this->authService->socialCallback(
            $provider,
            (string) $request->query('code', ''),
            $request->query('email'),
            $request->query('name'),
        ));
    }

    public function socialComplete(SocialCompleteRequest $request)
    {
        return $this->success($this->authService->completeSocialSignup(
            $request->user(),
            $request->string('phone')->value(),
            $request->string('nickname')->value(),
        ));
    }

    public function socialLink(SocialLinkRequest $request)
    {
        return $this->success($this->authService->linkSocial(
            $request->user(),
            $request->string('provider')->value(),
            $request->string('socialToken')->value(),
        ));
    }

    public function socialUnlink(Request $request, string $provider)
    {
        return $this->success($this->authService->unlinkSocial($request->user(), $provider));
    }
}

File: BackEnd/Java/springshop_maven/src/main/java/com/pbshop/springshop/auth/AuthController.java
package com.pbshop.springshop.auth;

import java.net.URI;
import java.util.Map;

import jakarta.validation.Valid;

import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.pbshop.springshop.auth.dto.AuthDtos.LoginRequest;
import com.pbshop.springshop.auth.dto.AuthDtos.PasswordResetConfirmRequest;
import com.pbshop.springshop.auth.dto.AuthDtos.PasswordResetRequest;
import com.pbshop.springshop.auth.dto.AuthDtos.PasswordResetVerifyRequest;
import com.pbshop.springshop.auth.dto.AuthDtos.RefreshRequest;
import com.pbshop.springshop.auth.dto.AuthDtos.ResendVerificationRequest;
import com.pbshop.springshop.auth.dto.AuthDtos.SignupRequest;
import com.pbshop.springshop.auth.dto.AuthDtos.SocialCompleteRequest;
import com.pbshop.springshop.auth.dto.AuthDtos.SocialLinkRequest;
import com.pbshop.springshop.auth.dto.AuthDtos.VerifyEmailRequest;
import com.pbshop.springshop.auth.security.AuthenticatedUserPrincipal;
import com.pbshop.springshop.common.api.ApiResponse;

@RestController
@RequestMapping("${pbshop.api.base-path:/api/v1}/auth")
public class AuthController {

    private final AuthService authService;

    public AuthController(AuthService authService) {
        this.authService = authService;
    }

    @PostMapping("/signup")
    public ApiResponse<Map<String, Object>> signup(@Valid @RequestBody SignupRequest request) {
        return ApiResponse.success(authService.signup(request));
    }

    @PostMapping("/verify-email")
    public ApiResponse<Map<String, Object>> verifyEmail(@Valid @RequestBody VerifyEmailRequest request) {
        return ApiResponse.success(authService.verifyEmail(request));
    }

    @PostMapping("/resend-verification")
    public ApiResponse<Map<String, Object>> resendVerification(@Valid @RequestBody ResendVerificationRequest request) {
        return ApiResponse.success(authService.resendVerification(request));
    }

    @PostMapping("/login")
    public ApiResponse<Map<String, Object>> login(@Valid @RequestBody LoginRequest request) {
        return ApiResponse.success(authService.login(request));
    }

    @PostMapping("/logout")
    public ApiResponse<Map<String, Object>> logout(@AuthenticationPrincipal AuthenticatedUserPrincipal principal) {
        return ApiResponse.success(authService.logout(principal));
    }

    @PostMapping("/refresh")
    public ApiResponse<Map<String, Object>> refresh(@Valid @RequestBody RefreshRequest request) {
        return ApiResponse.success(authService.refresh(request));
    }

    @PostMapping("/password-reset/request")
    public ApiResponse<Map<String, Object>> requestPasswordReset(@Valid @RequestBody PasswordResetRequest request) {
        return ApiResponse.success(authService.requestPasswordReset(request));
    }

    @PostMapping("/password-reset/verify")
    public ApiResponse<Map<String, Object>> verifyPasswordReset(@Valid @RequestBody PasswordResetVerifyRequest request) {
        return ApiResponse.success(authService.verifyPasswordReset(request));
    }

    @PostMapping("/password-reset/confirm")
    public ApiResponse<Map<String, Object>> confirmPasswordReset(@Valid @RequestBody PasswordResetConfirmRequest request) {
        return ApiResponse.success(authService.confirmPasswordReset(request));
    }

    @GetMapping("/login/{provider}")
    public ResponseEntity<Void> socialLogin(@PathVariable String provider) {
        URI location = authService.socialLoginRedirect(provider);
        return ResponseEntity.status(302).location(location).build();
    }

    @GetMapping("/callback/{provider}")
    public ApiResponse<Map<String, Object>> socialCallback(@PathVariable String provider, @RequestParam(required = false) String code) {
        return ApiResponse.success(authService.socialCallback(provider, code));
    }

    @PostMapping("/social/complete")
    public ApiResponse<Map<String, Object>> socialComplete(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @Valid @RequestBody SocialCompleteRequest request
    ) {
        return ApiResponse.success(authService.socialComplete(principal, request));
    }

    @PostMapping("/social/link")
    public ApiResponse<Map<String, Object>> socialLink(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @Valid @RequestBody SocialLinkRequest request
    ) {
        return ApiResponse.success(authService.socialLink(principal, request));
    }

    @DeleteMapping("/social/unlink/{provider}")
    public ApiResponse<Map<String, Object>> socialUnlink(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @PathVariable String provider
    ) {
        return ApiResponse.success(authService.socialUnlink(principal, provider));
    }
}
